PR: Eliminate unnecessary updates of the Variable Explorer while debugging

rear1019 opened the pull request:
See commit message for a detailed description. We have been using this patch for quite some time and haven’t experienced issues so far.

At least in theory, this should improve the performance of the debugger in Spyder (in terms of steps/second). In practical terms, performance gains are only noticeable with _very_ many variables in current scope. Think of about 80–100 variables on a rather old/slow machine, and > 200–300 variables on a rather fast one. Gains may be bigger with remote kernels, but I haven’t tested that

PS: Please let me know if I should provide instructions how to (roughly) measure the performance.

ccordoba12 commented:
@impact27, could you take a look at this one? Thanks!

impact27 commented:
Is `update_plugins` ever set to True? I can't see any instances in the PR.
If you add a variable with a command (simply `testvar = 1` in the debugger), is it shown in the variable explorer?

rear1019 commented:
> Is `update_plugins` ever set to True? I can't see any instances in the PR.

This is not required (also see below). I have added this parameter without a default value so that any user of `publish_pdb_state()` or `notify_spyder()` must explicitly check whether `update_plugins` must be `True` or `False`. However, I do see that this is confusing, so it’s possibly better to remove this parameter.

> If you add a variable with a command (simply `testvar = 1` in the debugger), is it shown in the variable explorer?

Yes, it is shown. Any modification of the current namespace is shown, just like before. I have tried to explain this patch in the lengthy commit message, but let’s give it another try:

The are two ways the Variables Explorer is updated:
* The kernel proactively sends updated data. This only happens inside the debugger.
* The Variable Explorer requests updated data from the kernel.

The former is not required due to the latter and is removed by this PR. The latter relies on the `executed` signal emitted by `QtConsole` [1]: The Variable Explorer requests an update of its data from the kernel upon the `executed` signal [2].

This signal is emitted when the kernel finishes execution of a command (in the non-debugger case). As an extension, this signal is also emitted by Spyder when the debugger requests input. The request is a remote call into Spyder, handled in `DebuggingWidget.pdb_input()` [2]. In any case, the `executed` signal signals that the kernel has finished execution of a command, implying that the namespace may have changed.

[1] https://github.com/jupyter/qtconsole/blob/5.3.0/qtconsole/frontend_widget.py#L145-L147
[2] https://github.com/spyder-ide/spyder/blob/v5.3.0/spyder/plugins/ipythonconsole/widgets/client.py#L427
[3] https://github.com/spyder-ide/spyder/blob/v5.3.0/spyder/plugins/ipythonconsole/widgets/debugging.py#L599

impact27 commented:
> The are two ways the Variables Explorer is updated:
> 
>     * The kernel proactively sends updated data. This only happens inside the debugger.
> 
>     * The Variable Explorer requests updated data from the kernel.
> 
> The former is not required due to the latter and is removed by this PR. 

This makes sense. Is there a reason why you elected to remove the former rather than the latter? Removing the latter instead would avoid an extra request wouldn't it?

Also the `executed` signal is only emitted when the execution is visible[1]. If Spyder does "hidden execution" there would be no request to update the variable explorer and it would therefore not be up to date? (To be fair I am not sure under which circumstances that would happen, maybe when updating a variable from the variable explorer?)

> This is not required (also see below). I have added this parameter without a default value so that any user of publish_pdb_state() or notify_spyder() must explicitly check whether update_plugins must be True or False. However, I do see that this is confusing, so it’s possibly better to remove this parameter.

I don't think it is needed to keep that if it is not used.


[1]https://github.com/spyder-ide/spyder/blob/c4cca83a68f9bd69ff2afb9d4396ea3d297bfdbc/spyder/plugins/ipythonconsole/widgets/debugging.py#L623

rear1019 commented:
> > The are two ways the Variables Explorer is updated:
> > ```
> > * The kernel proactively sends updated data. This only happens inside the debugger.
> > 
> > * The Variable Explorer requests updated data from the kernel.
> > ```
> > The former is not required due to the latter and is removed by this PR.
> 
> This makes sense. Is there a reason why you elected to remove the former rather than the latter? Removing the latter instead would avoid an extra request wouldn't it?

Yes, removing the latter (update requests by Spyder) and retaining the former (kernel proactively sends data) would avoid an extra request. This is something we considered, and we had a corresponding patch for the debugger-case (IIRC). IIRC, we opted not go this way because
* We already had a working patch (this PR).
* We didn’t have the time to find and verify a corresponding patch for the non-debugger case (“plain” IPython console).
* We didn’t want to fix only the debugger-case, leading a situation where the debugger and non-debugger case use different mechanisms. 

However, making this further improvement – kernel proactively sends data instead of Spyder requesting it – is still in our backlog :wink:.

> 
> Also the `executed` signal is only emitted when the execution is visible[1].
>
> [1]https://github.com/spyder-ide/spyder/blob/c4cca83a68f9bd69ff2afb9d4396ea3d297bfdbc/spyder/plugins/ipythonconsole/widgets/debugging.py#L623

Note: This also applies to the non-debugger case (“plain” IPython console) [1]

> If Spyder does "hidden execution" there would be no request to update the variable explorer and it would therefore not be up to date?

That’s correct, assuming the hidden execution modifies the current namespace (I hope not). However, this issue exists regardless whether this PR is applied or not.

To be honest, I didn’t looked too hard into hidden execution. I merely verified that debugger commands triggered by Spyder’s GUI are not hidden.

> (To be fair I am not sure under which circumstances that would happen, maybe when updating a variable from the variable explorer?)
> 

Changing a variable from within the Variable Explorer (as opposed to entering code in the IPython console) uses a different communication channel implemented by `spyder_kernels`/Spyder. No execution requests/replies (as defined by Jupyter and implemented by `ipykernel/qtconsole`) are involved and consequently no `executed` signal is emitted. That’s actually an issue for the Watchlist plugin I wasn’t aware of.

> > This is not required (also see below). I have added this parameter without a default value so that any user of publish_pdb_state() or notify_spyder() must explicitly check whether update_plugins must be True or False. However, I do see that this is confusing, so it’s possibly better to remove this parameter.
> 
> I don't think it is needed to keep that if it is not used.

I will remove it.

[1] https://github.com/jupyter/qtconsole/blob/5.3.0/qtconsole/frontend_widget.py#L449

impact27 commented:
> Note: This also applies to the non-debugger case (“plain” IPython console) [1]

This is a good point. I think we should keep the same code path for both debugger and non-debugger then.

I tried this PR and it breaks debugging following in spyder. If you use the up/down command the blue arrow doesn not follow the position of the debugger. I think you need to add back `postcmd`.

To test:

```python
test1 = 1
def a():
    test2 = 2
    return 1 + 1 # Breakpoint here
a()
```

1. debug file
2. use up command
3. See that the debug arrow is still on line 4

rear1019 commented:
Changes in force push:
* Fixed issue found by @impact27 [1]
* Removed `update_plugins` parameter

[1] https://github.com/spyder-ide/spyder-kernels/pull/381#issuecomment-1129987411

impact27 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -644,9 +644,14 @@ def precmd(self, line):
+        if line.startswith("!"):
+            line = line[1:]
+        cmd, arg, line = self.parseline(line.strip())
+        if cmd in ("up", "down", "jump", "u", "d", "j"):
```
It is a good idea to check wether the update is needed!

I am not sure I am a fan of the list of commands. This duplicates a part of the `line` processing and the list might be subject to change. Why not check if the frame has changed instead, which is what we want to know?

My proposition would be to move the logic inside `notify_spyder`, where the filename and line number can be compared to the previous update, saved in a variable (`self._previous_step` maybe?).
Then `self._previous_step = None` in `postloop` so the next `preloop` would send an update as expected.

Also maybe `notify_spyder` could be renamed now that it only sends the debugger position?

impact27 reviewed (changes requested):
Thanks for the work, I think this will help with the debugging speed!

impact27 commented on spyder_kernels/console/kernel.py:
```diff
@@ -311,11 +311,7 @@ def publish_pdb_state(self, step):
-            var_properties=self.get_var_properties(),
-            step=step
-         )
+        state = dict(step=step)
```
Now that `publish_pdb_state` doesn't use any kernel function I don't think it should stay in kernel. I would move the logic here in `notify_spyder`. 

You can replace `self.frontend_call(blocking=False)` by `frontend_request(blocking=False)`

rear1019 commented:
Thanks for the review @impact27! I have implemented your very sensible suggestions. Please let me now if I should rebase the patch against the most recent `master` or `2.x` branch.

impact27 commented:
Thanks! Looks good now!

@ccordoba12 what do you think about the branch?

ccordoba12 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -92,6 +93,9 @@ def __init__(self, completekey='tab', stdin=None, stdout=None,
         self._frontend_notified = False
 
+        # content of tuple: (filename, line number)
+        self._previous_step: Optional[Tuple[basestring, int]] = None
```
```suggestion
        self._previous_step = None
```

ccordoba12 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -13,6 +13,7 @@
 import sys
 import traceback
 from collections import namedtuple
+from typing import Tuple, Optional
```
```suggestion
```

This package still supports Python 2, so we need to remove this and the typing annotations you added below. We will migrate to Python 3 only in Spyder-kernels 3.

ccordoba12 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -542,6 +546,12 @@ def is_name_or_composed(text):
         return result
 
     # --- Methods overriden by us for Spyder integration
+    def postloop(self) -> None:
```
```suggestion
    def postloop(self):
```

ccordoba12 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -743,11 +754,13 @@ def set_spyder_breakpoints(self, breakpoints):
-    def notify_spyder(self):
-        """Send kernel state to the frontend."""
+    def publish_pdb_state(self):
+        """Send debugger state (frame position) to the frontend. The state is
+        only sent if it has changed since the last update."""
```
```suggestion
        """
        Send debugger state (frame position) to the frontend.

        The state is only sent if it has changed since the last update.
        """
```

Improve formatting of this docstring.

ccordoba12 reviewed (commented):
@rear1019, last review for you, then this should be ready.

rear1019 commented:
Force push:
* Fixed display of current code position (blue arrow in editor) after leaving a recursive debugger. @impact27 See the changes in `SpyderPdb.enter_recursive_debugger()`.
* Applied suggestions by @ccordoba12 to restore compatibility with Python 2
* Rebased against current `2.x` branch

impact27 commented:
> * See the changes in `SpyderPdb.enter_recursive_debugger()`.

Looks good thanks for checking that

impact27 commented on spyder_kernels/customize/spyderpdb.py:
```diff
@@ -821,6 +845,11 @@ def enter_recursive_debugger(self, code, filename,
+        # Notify spyder about changed frame. Reset _previous_step to force
+        # notification.
+        self._previous_step = None
+        self.publish_pdb_state()
```
I think this is not needed because `postcmd` will be called after this function returns. The line above (`self._previous_step = None`) should be sufficient to address the issue. Am I missing something?

rear1019 replied:
No, you’re not. Resetting `self._previous_step` is sufficient.

rear1019 commented:
Force push:
* Addressed comment [1]
* Added encoding line in `spyderpdb.py`. This should fix CI failures in with Python 2

[1] https://github.com/spyder-ide/spyder-kernels/pull/381#discussion_r897941057

ccordoba12 commented:
Thanks @rear1019 for the updates!

@impact27, what do you think about them?

impact27 commented:
Looks good to me 👍

ccordoba12 commented:
Great! Merging then.

ccordoba12 reviewed (approved):
Thanks @rear1019 for your work on this!
